skopt/plots.py

```diff
@@ -88,3 +89,65 @@ def plot_convergence(*args, **kwargs):
         ax.legend(loc="best")
 
     return ax
+
+
+def plot_scatter_matrix(result, bins=20):
+    """Plot scatter plot matrix of samples.
+
+    The diagonal shows a histogram for each dimension. Create pairwise
+    scatter plots of each dimension of the search space. Later
+    samples are shown in a darker color.
+
+    Parameters
+    ----------
+    * `result` [`OptimizeResult`]
+        The result for which to plot the scatter plot matrix.
+
+    * `bins` [int, bins=20]:
+        Number of bins to use for histograms on the diagonal.
+
+    Returns
+    -------
+    * `ax`: [`Axes`]:
+        The matplotlib axes.
+    """
+    samples = np.asarray(result.x_iters)
+    order = range(samples.shape[0])
+    fig, ax = plt.subplots(result.space.n_dims,
+                           result.space.n_dims,
+                           figsize=(8, 8))
+    fig.subplots_adjust(left=0.05, right=0.95, bottom=0.05, top=0.95,
+                        hspace=0.1, wspace=0.1)
+
+    for i in range(result.space.n_dims):
+        for j in range(i + 1):
+            if i == j:
+                ax[i, i].hist(samples[:, i], bins=bins)
```

Comment:
Independent of the colouring of the off-diagonals I think this is a good idea. I would add it as a line ontop of the histogram. I'd like to be able to see where the samples are (the histogram) and what the value of the objective function is (line). To verify that most samples are where the suspected minimum is.

Should we make a new function which shows on the diagonal the 1D partial dependence and on the off-diagonal the 2D?